### PR TITLE
File handling improvements

### PR DESCRIPTION
**Pull request by wotschofsky**

### CODE REVIEW

File: rm102/gui.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 from pathlib import Path
-from PyQt5.QtWidgets import QMainWindow, QApplication, QFileDialog, QDialog
+from PyQt5.QtWidgets import QMainWindow, QApplication, QFileDialog, QDialog, QMessageBox
```
Comment (bionade24) [inline rm102/gui.py]: QMessageBox on a new line would be better, to at least comply with "only 79 characters per line" PEP 8 style guideline  when it doesn't make the code less readable.

File: rm102/gui.py [hunk 2]
```diff
@@ -50,11 +51,30 @@ def update_gui(self, reg_count, acc, error):
...
+    def closeEvent(self, event):
+        if not self.changes_saved:
+            filename = filename = self.filepath.split('/')[-1] if self.filepath else 'Unbenannt'
+            msg = QMessageBox()
+            msg.setIcon(QMessageBox.Warning)
+            msg.setWindowTitle("Änderungen speichern?")
+            msg.setText(f"Sollen die Änderungen an {filename} gespeichert werden?")
+            msg.setStandardButtons(QMessageBox.Save | QMessageBox.Discard | QMessageBox.Cancel)
+            msg.setEscapeButton(QMessageBox.Cancel)
+            result = msg.exec()
```
Comment (bionade24) [inline rm102/gui.py:74]: I fear many lines here are too long, too :(

File: rm102/gui.py [hunk 3]
```diff
@@ -83,13 +103,14 @@ def loadButtonClicked(self):
             return
         with open(fname[0], 'r') as fobj:
             i = 0
-            for line in fobj.readlines():
+            lines = fobj.readlines()
+            for line in lines:
                 self.commandRegisterList.model.item(i).setText(line.strip('\n'))
                 i += 1
-                if self.commandRegisterList.model.item(i) is None:
+                if self.commandRegisterList.model.item(i) is None and i < len(lines):
```
Comment (bionade24) [inline rm102/gui.py]: Imho an empty line at the end of the CommandRegisterList a good and wanted thing.

File: rm102/gui.py [hunk 4]
```diff
@@ -50,11 +51,30 @@ def update_gui(self, reg_count, acc, error):
...
+            msg.setWindowTitle("Änderungen speichern?")
+            msg.setText(f"Sollen die Änderungen an {filename} gespeichert werden?")
+            msg.setStandardButtons(QMessageBox.Save | QMessageBox.Discard | QMessageBox.Cancel)
+            msg.setEscapeButton(QMessageBox.Cancel)
+            result = msg.exec()
+
+            if result == QMessageBox.Cancel:
+                event.ignore()
+            if result == QMessageBox.Save:
+                self.saveButtonClicked()
```
Comment (bionade24) [inline rm102/gui.py:79]: For the framework's convinience, there should be `event.accept()`

**Review by bionade24 — CHANGES_REQUESTED** [review]
Instead of this callback system, using Qt's signal/slot system, like as it is already used in the code, would be more appropriate. https://pythonspot.com/pyqt5-signals-and-slots/

File: rm102/gui.py (hunk 1 above)
Comment (wotschofsky) [reply rm102/gui.py]: Solved in 2fef9d64a201e1a54df39048c1c597cd0fe0d8ee

File: rm102/gui.py (hunk 2 above)
Comment (wotschofsky) [reply rm102/gui.py:74]: Solved in 2fef9d64a201e1a54df39048c1c597cd0fe0d8ee

File: rm102/gui.py (hunk 3 above)
Comment (wotschofsky) [reply rm102/gui.py]: Reverted in a02275279ec5adca4306b11867d0e5a32ebb5038

File: rm102/gui.py (hunk 4 above)
Comment (wotschofsky) [reply rm102/gui.py:79]: Solved in 4a8511e22ed1ab409cc91e0451d1582b76a42357

**Comment by bionade24** [comment]
I changed some things and squashed your commits into once to make it more clear as one feature addition and avoid Revert commits in the history.   
I also added detection of changes  in command registers and changed some minor stuff.

**Comment by bionade24** [comment]
Btw: While testing before the 0.3 release if found one really nasty bug: Pressing cancel on the quit on unsaved dialog closed the application. Fix is here: https://github.com/bionade24/registermachine102/commit/45e758f2b284bbdd94f4b22173935a8decd654f9 . I though you may want to know that.

**Pull request merged**